RCT/JX/get_swh_components_associated_with_each_swh_bat function (Re)

Jiarongx-Xie opened the pull request:
This function is ready for review.

Jiarongx-Xie commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py:
```diff
@@ -0,0 +1,108 @@
+        swh_and_equip_dict.setdefault(
+            swh_bat,
+            {
+                "swh_distribution": [],
```
If we use the dataclass created in 
![image](https://github.com/user-attachments/assets/4095cc98-65b0-4dcf-aa3b-900d03683da0), we need to exclude some existing fields and add a new field. In this case, do we still want to reuse it? Or do we create a new dataclass? @

Jiarongx-Xie commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py:
```diff
@@ -0,0 +1,108 @@
+                "solar_thermal": [],
+                "swh_uses": [],
+                "energy_required": ZERO.ENERGY,
+            },
```
If we use the dataclass created in [this merged fucntion](https://github.com/pnnl/ruleset-checking-tool/pull/1511/files)
![Screenshot 2024-10-30 230242](https://github.com/user-attachments/assets/eac0f618-cfdb-4bea-be5b-256ed331708a)
, we need to exclude some existing fields and add a new field. In this case, do we still want to reuse it? Or do we create a new dataclass? Do you have any recommendations? @weilixu

weilixu replied:
If changed, then create a new

Jiarongx-Xie replied:
Updated, thank you!

weilixu commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:
```diff
@@ -0,0 +1,123 @@
+            )
+
+            if swh_use.get("use_units") == SERVICE_WATERH_EATING_USE_UNIT.OTHER:
+                return RCTOutcomeLabel.UNDETERMINED
```
A function should not return UNDETERMINED outcome. I would suggest change this return value to None instead of a String or rule outcome.
Another convention is we do not return in the middle or logic - this is for the sake of maintenance.
Based on the description in the RDS, if any `swh_use` has `use_units` equal to "OTHER", return None.
We can do 
```python
if all([swh_use.get("use_units", SERVICE_WATERH_EATING_USE_UNIT.OTHER) == SERVICE_WATERH_EATING_USE_UNIT.OTHER for swh_use in service_water_heating_uses"]):
    ...
```

yunjoonjung-PNNL replied:
Hi Weili, In the comment from another PR, we decided to return ```UNDETERMINED``` instead of ```None```. Please let me know if we need to return ```None```.

weilixu commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:
```diff
@@ -0,0 +1,123 @@
+                        "id",
+                        space_id,
+                        building_segment,
+                    )
```
I think you can use the exactly one space - or create a new one.

yunjoonjung-PNNL replied:
Addressed.

weilixu commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:
```diff
@@ -0,0 +1,123 @@
+                        ] += swh_use_energy_by_space[space_id]
+                    else:
+                        swh_use_dict.setdefault(
+                            RCTOutcomeLabel.UNDETERMINED, ZERO.ENERGY
```
This is intermediate data - so you can put "UNDETERMINED" as the key - again, do not use RCTOutcomeLabel.

yunjoonjung-PNNL replied:
Addressed.

weilixu commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:
```diff
@@ -0,0 +1,123 @@
+        if swh_use_dict_len == 1 and RCTOutcomeLabel.UNDETERMINED not in swh_use_dict:
+            building_segment_swh_bat = next(iter(swh_use_dict))  # get the key
+        elif swh_use_dict_len == 1 and RCTOutcomeLabel.UNDETERMINED in swh_use_dict:
+            building_segment_swh_bat = RCTOutcomeLabel.UNDETERMINED
```
Set this scenario to None.

yunjoonjung-PNNL replied:
I addressed the comment from different PR. Addressed.

weilixu commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:
```diff
@@ -0,0 +1,123 @@
+                keys = list(swh_use_dict.keys())
+                values = list(swh_use_dict.values())
+
+                building_segment_swh_bat = keys[0] if values[0] > values[1] else keys[1]
```
My suggestion is do a ranking sort based on the value.
Then pick the largest as the building_segment_swh_bat. If the largest is UNDETERMINED, then return None.

weilixu reviewed (commented):
I think I did the same review on the building segment swh bat function previously and left pretty consistent comments.

yunjoonjung-PNNL commented:
@Jiarongx-Xie 

Great work! Because of my mistake with the ```get_swh_uses_associated_with_each_building_segment``` function, could you update the following code (from line 52-56) to the code shown below?

Current code
```python
service_water_heating_use_ids = (
            get_swh_uses_associated_with_each_building_segment(rmd)
        )
        for swh_use_id in service_water_heating_use_ids:
```

Suggested code
```python
for swh_use_id in find_all(
            f'$.buildings[*].building_segments[*][?(@.id="{building_segment["id"]}")].zones[*].spaces[*].service_water_heating_uses[*].id',
            rmd,
        ):
```

Jiarongx-Xie commented:
> @Jiarongx-Xie
> 
> Great work! Because of my mistake with the `get_swh_uses_associated_with_each_building_segment` function, could you update the following code (from line 52-56) to the code shown below?
> 
> Current code
> 
> ```python
> service_water_heating_use_ids = (
>             get_swh_uses_associated_with_each_building_segment(rmd)
>         )
>         for swh_use_id in service_water_heating_use_ids:
> ```
> 
> Suggested code
> 
> ```python
> for swh_use_id in find_all(
>             f'$.buildings[*].building_segments[*][?(@.id="{building_segment["id"]}")].zones[*].spaces[*].service_water_heating_uses[*].id',
>             rmd,
>         ):
> ```

Revised, thank you for the reminder.

Jiarongx-Xie commented:
Comments on the [get_building_segment_swh_bat](https://github.com/pnnl/ruleset-checking-tool/pull/1517) are addressed in the related PR. I resolved these comments without addressing them within this PR. @weilixu

yunjoonjung-PNNL commented:
Hi @Jiarongx-Xie 

Weili left some comments for this function in PR #1543 and I addressed them. Can you compare the function in this PR and PR #1543 and finalize this function development?

yunjoonjung-PNNL commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:
```diff
@@ -0,0 +1,123 @@
+        "id",
+        building_segment_id,
+        rmd,
+    )
```
Addressed.

yunjoonjung-PNNL commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:
```diff
@@ -0,0 +1,123 @@
+        building_segment_swh_bat = building_segment[
+            "service_water_heating_building_area_type"
+        ]
+    else:
```
Addressed.

yunjoonjung-PNNL commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:
```diff
@@ -0,0 +1,123 @@
+                "id",
+                swh_use_id,
+                building_segment,
+            )
```
Right! Addressed.

weilixu commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py:
```diff
@@ -0,0 +1,106 @@
+    for building_segment in find_all("$.buildings[*].building_segments[*]", rmd):
+        swh_bat = get_building_segment_swh_bat(rmd, building_segment["id"])
+        swh_and_equip_dict[swh_bat] = SWHEquipmentAssociations(
+            energy_required=ZERO.ENERGY
```
Do we need this? I am not sure since the energy_required default is ZERO.ENERGY

Jiarongx-Xie replied:
Yes, it gave` parameter "energy_required" unfilled `error if removing this.

weilixu commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py:
```diff
@@ -0,0 +1,106 @@
+            swh_use = find_exactly_one_service_water_heating_use(rmd, swh_use_id)
+            energy_required = get_energy_required_to_heat_swh_use(
+                swh_use_id, rmd, building_segment["id"], is_leap_year
+            )
```
I think we can loop the `swh_use` object and take out the `id` from the `swh_use` object.

weilixu commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py:
```diff
@@ -0,0 +1,106 @@
+            f'$.buildings[*].building_segments[*][?(@.id="{building_segment["id"]}")].zones[*].spaces['
+            f"*].service_water_heating_uses[*].id",
+            rmd,
+        ):
```
Also, add `TODO` comment here in case of schema change

Jiarongx-Xie replied:
Added

Jiarongx-Xie replied:
Done

weilixu commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py:
```diff
@@ -0,0 +1,106 @@
+            if list(energy_required.values())[0] is not None:
+                swh_and_equip_dict[swh_bat].energy_required += list(
+                    energy_required.values()
+                )[0]
```
Why only picking out the index 0?
Shouldnt this be the full set of energy-spaces in the `energy_required`?

Jiarongx-Xie replied:
Good point, addressed!

Jiarongx-Xie replied:
Good catch! Addressed

weilixu reviewed (approved):
Approved but have four comments need to address

Jiarongx-Xie commented:
> @Jiarongx-Xie
> 
> Great work! Because of my mistake with the `get_swh_uses_associated_with_each_building_segment` function, could you update the following code (from line 52-56) to the code shown below?
> 
> Current code
> 
> ```python
> service_water_heating_use_ids = (
>             get_swh_uses_associated_with_each_building_segment(rmd)
>         )
>         for swh_use_id in service_water_heating_use_ids:
> ```
> 
> Suggested code
> 
> ```python
> for swh_use_id in find_all(
>             f'$.buildings[*].building_segments[*][?(@.id="{building_segment["id"]}")].zones[*].spaces[*].service_water_heating_uses[*].id',
>             rmd,
>         ):
> ```

addressed

yunjoonjung-PNNL commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:
```diff
@@ -0,0 +1,123 @@
+
+    """
+
+    building_segment = find_exactly_one_with_field_value(
```
In the latest PR, I added a ```find_exactly_one_building_segment``` utility function. Can you use it to simplify the code?

Jiarongx-Xie replied:
Thank you for the reminder! This method is not needed after I updated the implementation.

yunjoonjung-PNNL commented on rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:
```diff
@@ -0,0 +1,123 @@
+    else:
+        swh_use_dict = {}
+        service_water_heating_use_ids = (
+            get_swh_uses_associated_with_each_building_segment(rmd, building_segment_id)
```
Yes. Addressed.

This part can now be changed to the following.
```Python
 service_water_heating_use_ids  = get_swh_uses_associated_with_each_building_segment(rmd)["building_segment_id"]
```

yunjoonjung-PNNL reviewed (approved):
Great work. I left two stylish comments. Thank you for this effort Jiarong!

Jiarongx-Xie commented:
> I think I did the same review on the building segment swh bat function previously and left pretty consistent comments.

I resolved all comments on [get_building_segment_swh_bat.py](https://github.com/pnnl/ruleset-checking-tool/pull/1548/files/17d8c61c23997c60e1ad7df14435cfe5157aaece#diff-64847523569d84f6c199ce193cc80ed4e6af557f20487beaa415ef3739727a55). Yun Joon has addressed these comments in that PR.

Jiarongx-Xie commented:
> Great work. I left two stylish comments. Thank you for this effort Jiarong!

thanks, Yun Joon!
